### PR TITLE
Adds TestEndpointReachable (for yaml test-cases)

### PR DESCRIPTION
**Pull request by aledsage**

### CODE REVIEW

File: test-framework/src/main/java/org/apache/brooklyn/test/framework/TestEndpointReachableImpl.java [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+            throw new IllegalArgumentException(String.format("The entity [%s] with endpoint [%s] has no port", getClass().getName(), endpoint));
+        }
+        return result;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public void stop() {
+        ServiceStateLogic.setExpectedState(this, Lifecycle.STOPPING);
```
Comment (neykov) [inline test-framework/src/main/java/org/apache/brooklyn/test/framework/TestEndpointReachableImpl.java]: Should be `STOPPED`

File: test-framework/src/test/java/org/apache/brooklyn/test/framework/TestEndpointReachableTest.java [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+    }
+
+    protected ServerSocket openServerPort() {
+        int startPort = 58767;
+        int endPort = 60000;
+        int port = startPort;
+        InetAddress localAddress = Networking.getLocalHost();
+        do {
+            try {
+                ServerSocket ss = new ServerSocket(port, 1024, localAddress);
```
Comment (neykov) [inline test-framework/src/test/java/org/apache/brooklyn/test/framework/TestEndpointReachableTest.java]: Passing port `0` will let the OS allocate an available port - no need for the retry logic.

**Comment by neykov** [comment]
Looks good, nice test coverage. What do you think about adding an option to reverse the test - successful if can't connect (for example by adding `assertFails: true`). It's as useful doing the reverse check.
Current implementation of `Networking.isReachable` might take a while to give up if the port is not reachable.

File: test-framework/src/test/java/org/apache/brooklyn/test/framework/TestEndpointReachableTest.java (hunk 2 above)
Comment (aledsage) [inline test-framework/src/test/java/org/apache/brooklyn/test/framework/TestEndpointReachableTest.java]: Ah, doh: you're absolutely right!

**Comment by aledsage** [comment]
I also wondered about the inverse. Agree it's worth adding. For consistency with the other TestCase types, I'll go for YAML syntax like that below (where it defaults to `reachable: true`):

```
- type: org.apache.brooklyn.test.framework.TestEndpointReachable
  name: Endpoint reachable
  brooklyn.config:
    targetId: entity-under-test
    endpointSensor: datastore.url
    assertions:
      reachable: false
```

I'm not going to worry about `Networking.isReachable` being slow to fail just now.

**Comment by neykov** [comment]
+1 for proposed syntax

**Comment by neykov** [comment]
Good to merge, once Jenkins is happy. 
Current failure: `[ERROR] TestEndpointReachableImpl.java[47:1] (imports) IllegalImport: Import from illegal package - com.google.api.client.repackaged.com.google.common.base.Objects.`

**Comment by aledsage** [comment]
Thanks @neykov - merging now.

**Pull request merged**